Basic deserialization.

beauby opened the pull request:
# Latest update:
Recap on the current situation:
+ Two methods are exposed (`ActiveModelSerializers::Deserialization.jsonapi_parse` and `ActiveModelSerializers::Deserialization.jsonapi_parse!`)
+ They both take the same parameters:
  - a `document` (either a Hash or an instance of `ActionController::Parameters` representing a JSON API payload)
  - an `options` hash:
    - `only` or `except`: array of white/black-listed fields (recall that in the context of JSON API, "fields" denotes both attributes and relationships)
    - `polymorphic`: array of polymorphic associations (those for which the `#{relationship}_type` must be set for `ActiveRecord` to handle it properly)
    - `keys`: hash of translated keys (for instance when the payload contains an `author` relationship which refers to an `user` association on the model)
+ The "unsafe" method (the one ending with a `!`) throws an `InvalidDocument` exception when the payload does not meet basic criteria for JSON API deserialization.
+ The "safe" method simply returns an empty hash.

Example:
```ruby
def create
  @post = Post.new(jsonapi_create_params)
end

def jsonapi_create_params
  ActiveModelSerializers::Deserialization.jsonapi_parse(params, only: [:title, :content])
end
```

~~There should be a bit more tests, but other than that,~~ I believe this PR is ready. Any thoughts?

# Update:
This PR offers basic JSON API deserialization through `ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(hash, options = {})` which takes a hash ~~or an instance of `ActionController::Parameters`~~ representing a JSON API document (and possibly some options, more on that further down) and returns an `ActiveRecord`-friendly hash, so that you can use it as follows:
```ruby
post_params = ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(params.to_h) # or params.to_unsafe_h depending on your rails version
@post = Post.new(post_params)
```

The only available option currently is `fields`, which allows for
 + whitelisting of fields (attributes and relationships) by providing an array of symbols, and 
 + specifying the corresponding attribute name on the model when using the `key` option on the association in the serializer, by providing a Hash at the end of the array.
Example:
```ruby
hash = {
  'data' => {
    'type' => 'photos',
    'id' => 'zorglub',
    'attributes' => {
      'title' => 'Ember Hamster',
      'src' => 'http://example.com/images/productivity.png'
    },
    'relationships' => {
      'author' => {
        'data' => nil
      },
      'photographer' => {
        'data' => { 'type' => 'people', 'id' => '9' }
      },
      'comments' => {
        'data' => [
          { 'type' => 'comments', 'id' => '1' },
          { 'type' => 'comments', 'id' => '2' }
        ]
      }
    }
  }
}

parse(hash, fields: [:id, :title, :src, :comments, author: :user])
```
returns
```ruby
{
  id: 'zorglub',
  src: 'http://example.com/images/productivity.png',
  user_id: nil,
  comment_ids: ['1', '2']
}
```

Note: It currently does not handle polymorphic associations, but could easily be extended to do so.

# Initial post:

This PR offers basic JSON API deserialization through ~~`ActiveModel::Serializer::Adapter::JsonApi.parse(params_or_hash)`~~ which takes a hash or an instance of `ActionController::Parameters` representing a JSON API document and turns it into an `ActiveRecord`-friendly hash, so that you cas use it as follows:
```ruby
post_params = ActiveModel::Serializer::Adapter::JsonApi.parse(params)
@post = Post.new(post_params)
```

~~Note that this PR does not support aliased relationships~~ (that is the name of the relationship in the client-issued JSON API document won't be changed). 
Moreover, ~~it currently does not support any kind of whitelisting directly~~, although you could simply do `post_params.slice!(:id, :name, :description)` to whitelist or `post_params.except!(:is_admin)` to blacklist.
Finally, ~~it currently fails if the JSON API document is ill-formed.~~

We could also make it so that the `parse` method transforms whatever comes its way into a new `ActionController::Parameters` instance, so that people can `require` / `permit` stuff instead of `slice` / `except`.

NullVoxPopuli commented on lib/active_model/serializer/adapter/json_api.rb:
```diff
@@ -41,6 +41,38 @@ def object
+        def self.parse(document)
+          hash = {}
+
+          hash[:id] = document['data']['id'] if document['data']['id']
```
should there be a check for the existence of `'data'`? could be handy to have detailed parse errors

beauby replied:
Sure. C.f. the description of the PR, there is currently no error handling, but it is planned.

NullVoxPopuli replied:
:+1:

bf4 replied:
Can we try to stick to the ActiveModel::Serializers::Json API
https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/serializers/json.rb#L137
`from_json`?

For completeness I'd also like to compare this to other libraries out there
and what language they use.

I think, as we've discussed, there should be a DeserializableDocument class
that has a from_json method on it, and that perhaps uses a Deserializer
underneath ref:
https://github.com/rails-api/active_model_serializers/issues/1098

beauby replied:
Rails' `from_json` and this method do different stuff. Here we are not building an instance of a model, we are just translating a json payload into a hash that can be passed to an AR model's constructor.

beauby commented:
Updated with handling of ill-formed documents, whitelisting, and aliased relationships.

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,73 @@
+          #      specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
+          def parse(document, options = {})
```
[`from_json`?](https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/serializers/json.rb#L137-L141)

beauby replied:
The issue with `from_json` is that it should be called on a model. Here, we make no assumption about the model. I guess we could override `ActiveModel::Serializers::JSON` so that it calls the `parse` method and updates the model. But it would be JSON API specific. Thoughts?

bf4 replied:
I just meant as a consistent interface

beauby replied:
My question still holds though. As a consistent interface of what? to what?

NullVoxPopuli replied:
I think calling from_json on json-api data should generate a model / associations / whatever.

so, imo, from_json could call this method, `parse` and return an instance of a model.

bf4 commented on test/adapter/json_api/parse_test.rb:
```diff
@@ -0,0 +1,73 @@
+                src: 'http://example.com/images/productivity.png',
+                author_id: nil,
+                photographer_id: '9',
+                comment_ids: %w(1 2)
```
Does Rails use comment_ids on a has_many :comments to infer the relationship via foreign key?

beauby replied:
Yes, c.f. http://guides.rubyonrails.org/association_basics.html section 4.3.1.

sachse commented:
Posting something like `{"data":{"attributes":{"email":"john@doe.dev","password":"qwerty"}}}` I get an error: 
```undefined method `each_with_object' for {"email"=>"john@doe.dev", "password"=>"qwerty"}:ActionController::Parameters```

Is this on my end?

beauby commented:
@sachse Your payload looks ok. Could you give me your ruby version, rails version, platform and a full stacktrace?
Also, if you have time, could you clone this branch and run the tests (`$ rake test`)?

sachse commented:
@beauby I'm running Ruby 2.2.3 and Rails 5 from master on OS X. The test suite passes without any problems.

Full trace:
```
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/active_model_serializers-a8564dcfe8f0/lib/active_model/serializer/adapter/json_api/deserialization.rb:46:in `parse_attributes'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/active_model_serializers-a8564dcfe8f0/lib/active_model/serializer/adapter/json_api/deserialization.rb:25:in `parse'
app/controllers/v1/sessions_controller.rb:18:in `session_params'
app/controllers/v1/sessions_controller.rb:13:in `create'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal/basic_implicit_render.rb:4:in `send_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/abstract_controller/base.rb:183:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal/rendering.rb:30:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/callbacks.rb:97:in `__run_callbacks__'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/callbacks.rb:743:in `_run_process_action_callbacks'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/callbacks.rb:90:in `run_callbacks'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/abstract_controller/callbacks.rb:19:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal/rescue.rb:29:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal/instrumentation.rb:31:in `block in process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/notifications.rb:164:in `block in instrument'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/notifications.rb:164:in `instrument'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal/instrumentation.rb:29:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal/params_wrapper.rb:248:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activerecord/lib/active_record/railties/controller_runtime.rb:18:in `process_action'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/abstract_controller/base.rb:128:in `process'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal.rb:192:in `dispatch'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_controller/metal.rb:264:in `dispatch'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/routing/route_set.rb:48:in `dispatch'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/routing/route_set.rb:32:in `serve'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/journey/router.rb:42:in `block in serve'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/journey/router.rb:29:in `each'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/journey/router.rb:29:in `serve'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/routing/route_set.rb:717:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionview/lib/action_view/digestor.rb:14:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rack-35599cfc2751/lib/rack/etag.rb:25:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rack-35599cfc2751/lib/rack/conditional_get.rb:38:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rack-35599cfc2751/lib/rack/head.rb:12:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activerecord/lib/active_record/query_cache.rb:36:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:963:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activerecord/lib/active_record/migration.rb:489:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/callbacks.rb:97:in `__run_callbacks__'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/callbacks.rb:743:in `_run_call_callbacks'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/callbacks.rb:90:in `run_callbacks'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/reloader.rb:71:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:48:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/railties/lib/rails/rack/logger.rb:42:in `call_app'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/railties/lib/rails/rack/logger.rb:24:in `block in call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/tagged_logging.rb:70:in `block in tagged'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/tagged_logging.rb:26:in `tagged'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/tagged_logging.rb:70:in `tagged'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/railties/lib/rails/rack/logger.rb:24:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/request_id.rb:24:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rack-35599cfc2751/lib/rack/runtime.rb:22:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/activesupport/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/load_interlock.rb:13:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/actionpack/lib/action_dispatch/middleware/static.rb:133:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rack-35599cfc2751/lib/rack/sendfile.rb:111:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rails-09463183867f/railties/lib/rails/engine.rb:522:in `call'
/Users/kristoffer/.gem/ruby/2.2.3/bundler/gems/rack-35599cfc2751/lib/rack/handler/webrick.rb:86:in `service'
/Users/kristoffer/.rubies/ruby-2.2.3/lib/ruby/2.2.0/webrick/httpserver.rb:138:in `service'
/Users/kristoffer/.rubies/ruby-2.2.3/lib/ruby/2.2.0/webrick/httpserver.rb:94:in `run'
/Users/kristoffer/.rubies/ruby-2.2.3/lib/ruby/2.2.0/webrick/server.rb:294:in `block in start_thread'
```

peterhicks commented:
I'm getting the same issue. I'm running Ruby 2.2.3, Rails 5 (from master), and AMS from your fork as of 10 minutes ago.

With the following code:

`@post_params = ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(params)`
`@genre = Genre.new(genre_post_params)`

I get this stack trace:
```
ActionController::RoutingError at /genres
=========================================

> undefined method `each_with_object' for {"name"=>"test"}:ActionController::Parameters

/Users/peter/Sites/rails/actionpack/lib/action_dispatch/routing/route_set.rb, line 35
-------------------------------------------------------------------------------------

``` ruby
   30             controller = controller req
   31             res        = controller.make_response! req
   32             dispatch(controller, params[:action], req, res)
   33           rescue NameError => e
   34             if @raise_on_name_error
>  35               raise ActionController::RoutingError, e.message, e.backtrace
   36             else
   37               return [404, {'X-Cascade' => 'pass'}, []]
   38             end
   39           end
   40   
```

App backtrace
-------------



Full backtrace
--------------

 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/routing/route_set.rb:35:in `rescue in serve'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/routing/route_set.rb:29:in `serve'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/journey/router.rb:42:in `block in serve'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/journey/router.rb:29:in `serve'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/routing/route_set.rb:717:in `call'
 - bullet (4.14.10) lib/bullet/rack.rb:10:in `call'
 - /Users/peter/Sites/rails/actionview/lib/action_view/digestor.rb:14:in `call'
 -  () Users/peter/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/rack-35599cfc2751/lib/rack/etag.rb:25:in `call'
 -  () Users/peter/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/rack-35599cfc2751/lib/rack/conditional_get.rb:38:in `call'
 -  () Users/peter/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/rack-35599cfc2751/lib/rack/head.rb:12:in `call'
 - /Users/peter/Sites/rails/activerecord/lib/active_record/query_cache.rb:36:in `call'
 - /Users/peter/Sites/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:963:in `call'
 - /Users/peter/Sites/rails/activerecord/lib/active_record/migration.rb:489:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/callbacks.rb:97:in `__run_callbacks__'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/callbacks.rb:743:in `_run_call_callbacks'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/callbacks.rb:90:in `run_callbacks'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/reloader.rb:71:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
 - better_errors (2.1.1) lib/better_errors/middleware.rb:84:in `protected_app_call'
 - better_errors (2.1.1) lib/better_errors/middleware.rb:79:in `better_errors_call'
 - better_errors (2.1.1) lib/better_errors/middleware.rb:57:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:48:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
 - /Users/peter/Sites/rails/railties/lib/rails/rack/logger.rb:42:in `call_app'
 - /Users/peter/Sites/rails/railties/lib/rails/rack/logger.rb:24:in `block in call'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/tagged_logging.rb:70:in `block in tagged'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/tagged_logging.rb:26:in `tagged'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/tagged_logging.rb:70:in `tagged'
 - /Users/peter/Sites/rails/railties/lib/rails/rack/logger.rb:24:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/request_id.rb:24:in `call'
 -  () Users/peter/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/rack-35599cfc2751/lib/rack/runtime.rb:22:in `call'
 - /Users/peter/Sites/rails/activesupport/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/load_interlock.rb:13:in `call'
 - /Users/peter/Sites/rails/actionpack/lib/action_dispatch/middleware/static.rb:133:in `call'
 -  () Users/peter/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/rack-35599cfc2751/lib/rack/sendfile.rb:111:in `call'
 - /Users/peter/Sites/rails/railties/lib/rails/engine.rb:522:in `call'
 -  () Users/peter/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/rack-35599cfc2751/lib/rack/handler/webrick.rb:86:in `service'
 - /Users/peter/.rbenv/versions/2.2.3/lib/ruby/2.2.0/webrick/httpserver.rb:138:in `service'
 - /Users/peter/.rbenv/versions/2.2.3/lib/ruby/2.2.0/webrick/httpserver.rb:94:in `run'
 - /Users/peter/.rbenv/versions/2.2.3/lib/ruby/2.2.0/webrick/server.rb:294:in `block in start_thread'
```

beauby commented:
Thanks @sachse. I believe ActionController::Parameters changed a bit in Rails5. I'll find a workaround tonight.

beauby commented:
Couldn't resist to see what's going on. Since [this commit](https://github.com/rails/rails/commit/14a3bd520dd4bbf1247fd3e0071b59c02c115ce0) `ActionController::Parameters` do not inherit from `HashWithIndifferentAccess` anymore, hence the issue at hand. Pushing a fix in 2mins.

jmonteiro commented on test/adapter/json_api/parse_test.rb:
```diff
@@ -0,0 +1,73 @@
+
+            def test_parameters
+              parameters = ActionController::Parameters.new(@hash)
+              parsed_hash = ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(parameters)
```
One solution to fixing this would be to send `parameters.to_unsafe_h` instead of only `parameters`. Or to try to call that method before calling `to_h` inside the `#parse` method.

beauby replied:
@jmonteiro The purpose of this test is to ensure the `parse` method can handle `ActionController::Parameters` in order to reduce friction for the user.

beauby replied:
The idea is that deserialization should handle parameters transparently. The preferred way for whitelisting would be through the `parse` method, but in case a user wants to do whitelisting on the  `ActionController::Parameters` instance, we should not override its will, that's why we're using `to_h` instead of `to_unsafe_h` inside `parse`.

beauby replied:
Oh, I may have made a false assumption here. I thought a pristine instance of `ActionController::Parameters` would have all its parameters whitelisted by default.

beauby replied:
Ok the catch here is that when initializing an instance of `ActionController::Parameters` like here, it will automatically whitelist all attributes, hence the test wouldn't catch the problem you're mentioning. However, the `to_unsafe_h` method was only added in rails 4.2, so I'm not sure what the best solution is here.

beauby replied:
For now, I'll remove direct support for `ActionController::Parameters`.

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,72 @@
+          #   fields: Array of symbols and a Hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
```
`@example` please.. needs more docs :)

NullVoxPopuli replied:
:+1:

bf4 replied:
Is this the flow?

1. deserialize_document
  1. deserialize_data
    1. deserialize_resource
      1. deserialize_relationships
  1. deserialize_meta
  1. deserialize_included
  1. deserialize_links

NullVoxPopuli replied:
This should be in a method header comment on the parse method (or something like it)

beauby replied:
except there is no ii. or iv.

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,72 @@
+            hash
+          end
+
+          # @api private
```
this converts options[:fields] into nil or a Hash?  Would `extract_options` work?

beauby replied:
yes and nope, respectively

NullVoxPopuli replied:
@beauby looks like you need to add some method header comments :-\

bf4 replied:
```ruby
# Given an Array
# For each element in the array, built a hash
# Then a symbol element is is used as both the key and value
# And a Hash element is merged into return value
# And all other values in the array are ignored
# @example
#    parse_fields_options(['a']) # => nil
#    parse_fields_options([:a]) # => { a: :a }
#    parse_fields_options([{'a' => :a]) # => { 'a' => :a }
#    parse_fields_options(nil) # => nil
#    parse_fields_options({}) # => nil
#    parse_fields_options(:a) # => nil
#    parse_fields_options([[:a]]) # => {}
#    parse_fields_options([:a, {a: :b}]) # => { a: :b }
#    parse_fields_options([{a: :b}, :a]) # => { a: :a }
#   etc
#   etc
# @params fields [Array<Hash, Symbol>] object to transform into a hash or nil
# @return [Hash] for array inputs
# @return [Nil] otherwise
```

GCorbel commented:
It works great. It should be merged! Thanks for your work.

beauby commented:
Rebased. Any feedback on this? I believe it is pretty useful.

bf4 commented:
Any updates since last discussion? Also, ref: https://github.com/rails-api/active_model_serializers/issues/1098

NullVoxPopuli commented:
my only feedback was for inline documentation, which I know @beauby despises... so...

:-)

NullVoxPopuli commented:
but other than that, I think it should be merged -- just to get things moving.

:+1:

bf4 commented on test/adapter/json_api/parse_test.rb:
```diff
@@ -0,0 +1,67 @@
+
+            def test_illformed_payload
+              parsed_hash = ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse({})
+              assert_equal({}, parsed_hash)
```
I don't understand what this is testing.  What would a failure look like?  (also, `malformed` or `invalid`?)

beauby replied:
This test just makes sure AMS does not go crazy when the hash does not comply to the JSON API spec. I guess ideally there should be a bunch more tests for invalid/missing keys.

bf4 commented on test/adapter/json_api/parse_test.rb:
```diff
@@ -0,0 +1,67 @@
+              @hash = {
+                'data' => {
+                  'type' => 'photos',
+                  'id' => 'zorglub',
```
So, this only works on single-resource documents?

beauby replied:
Yes, as per JSON API spec.

bf4 replied:
the spec doesn't allow deserializing collections?

beauby replied:
Nope. C.f. [Creating Resources](http://jsonapi.org/format/#crud-creating):
>  The request MUST include a single resource object as primary data.

[Updating Resources](http://jsonapi.org/format/#crud-updating):
> The PATCH request MUST include a single resource object as primary data.

bf4 replied:
But that only means the server expects single-resource requests (without the bulk extension)

The client would still need to handle multiple-resource responses

bf4 replied:
which I suppose is fine to start out only deserializing requests since that's the most common use-case

beauby replied:
Not sure I understand what you mean. Could you elaborate? Without the bulk extension, the only payloads a server will ever deal with are single-resource ones.

bf4 replied:
Not sure if you replied before I wrote https://github.com/rails-api/active_model_serializers/pull/1248/files#r46101704

beauby replied:
Oh, right. Yeah, this PR was initially intended to solve request deserialization only. Note that it would be fairly easy to build upon the `parse` method to parse collections.

bf4 commented on test/adapter/json_api/parse_test.rb:
```diff
@@ -0,0 +1,67 @@
+      class JsonApi
+        module Deserialization
+          class ParseTest < Minitest::Test
+            def setup
```
Needs a 'reversible' test.

1. Given a resource
2. Serialize resource to hash
3. deserialize to resource (attributes?)

bf4 replied:
and an integration test as described in https://github.com/rails-api/active_model_serializers/pull/1248/files#r46101225

beauby replied:
Added an integration test. However, after giving it some thoughts, I don't see the relevance of a 'reversible' test. What we are doing here is precisely transforming a JSON API-compliant document into an `ActiveRecord::Attributes`-compliant hash, so I'm not sure what introducing a dependency to the serialization would add.

bf4 replied:
>  introducing a dependency to the serialization would add

I mean like

1. Given: `{ 'data' : { 'posts' : [ { 'id' : '1' } ] } }`
1. Deserialize to `[ Post.new(id: 1) ]`
1. Serialize to `{ 'data' : { 'posts' : [ { 'id' : '1' } ] } }`

beauby replied:
I kinda disagree. The job of this piece of code is to do a transformation between two specified formats (JSON API on the one side, ActiveRecord#initialize on the other), which is currently being tested. I'm not against an integration test as described in your scenario, though. It just does not seem necessary to me to ensure the validity of this code.

joaomdmoura replied:
I agree that this is not the topic here, but would be a great test to have :smile:

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,72 @@
+        module Deserialization
+          module_function
+
+          # Parse a Hash representing a JSON API document into an ActiveRecord-ready hash.
```
Maybe `Transform a JSON API document, containing a single data object, from a hash into an ActiveRecord#attributes hash`?

beauby replied:
Looks good, will update.

beauby commented:
Added an alias from `ActiveModelSerializers::Deserialization.jsonapi_parse` to `ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse`.

This PR now seems ready for merge in my opinion. Thoughts @rails-api/ams?

NullVoxPopuli commented:
LGTM :-)

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,115 @@
+          end
+
+          # @api private
+          def parse_attributes(attributes, fields)
```
?

1. Given fields and truthy attributes
  1. Build a hash from each key/value pair
    1. If fields is truthy
      1. And there is a value for the symbolized key
         1. Add that value as the key for the original hash's value
     1.  And the key can be symbolized
        1. Add that symbolized key as the key for the original hash's value
1. Else `{}`

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,115 @@
+          end
+
+          # @api private
+          def parse_relationships(relationships, fields)
```
?

1. Given fields and truthy relationships
  1. Build a hash from each key/value pair
    1. Derive the `association_name`
      1. If fields is truthy
        1. And there is a value for the symbolized key
         1. Use that value
       1.  And the key can be symbolized
          1. Use the symbolized key
    1. Get the value of `data` from the value
      1. If `data` is an array (of hashes)
        1. Derive the key as `"#{association_name.to_s.singularize.to_sym}_ids".to_sym`
        1. Add the mapped `id` field of each element in `data` as the value to the key
      1. else
        1. derive the key as `#{association_name}_id".to_sym`
        1. if `data` is truthy
          1. set the value of data's `id` field as the value to the key
        1. else
          1. set `nil` as the value to the key

    
1. Else `{}`

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,115 @@
+          module_function
+
+          # Transform a JSON API document, containing a single data object, from a hash
+          # into an ActiveRecord#attributes hash.
```
What does ` into an ActiveRecord#attributes hash` mean?

bf4 commented:
I like what this does, but worry that it will be hard to maintain since it's kind of hard to follow what it's doing.  Do my comments about what each method does make sense?  Could we make use of `ActiveModel::Serializers::JSON` to just deserialize to that interface so that there's less to do?

frank06 commented:
Hi! what's the status of deserialization?  Any plans of releasing 0.10 soon?

beauby commented:
@frank06 Deserialization (in the sense of transforming a hash/ActionController::Parameters instance into a hash that is ready for `Model.new()` and such) is ready and working (you can use this branch for now). I'd like to get it merged ASAP, but as you can see there is a tiny bit of documenting work left before we can do that. As far as 0.10 is concerned, we're a bit late on the schedule but it should be ready quite soon.

frank06 commented:
Thanks @beauby for the heads up! People in the Ember community choose AMS and prefer JSON API... so it would be great if I could start recommending a fully working version

beauby commented:
Update: 
 + two flavors of `parse` now available (either via either `ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse`/``ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse!` or `ActiveModelSerializers::Deserialization.jsonapi_parse`/`ActiveModelSerializers::Deserialization.jsonapi_parse!`), one of which raises `ArugmentError 'Invalid payload'` on invalid payloads, and the other silently fails and returns an empty hash.
 + added tests
 + added yarddocs

beauby commented:
@bf4: `ActiveModel::Serializers::JSON.from_json` does something totally different, it builds a model from a plain json representation.

beauby commented:
@bf4: Moreover, I believe what's happening here is quite straightforward, the complexity is inherent to the JSON API format.

bf4 commented:
> Moreover, I believe what's happening here is quite straightforward, the complexity is inherent to the JSON API format

@beauby You make me feel bad when you say it's 'quite straightforward', as if there's something wrong with me that I asked.

I'll review the doc updates later.  I see it still refers to active record attributes ... Doesn't look ar specific to me.

beauby commented:
@bf4 Sorry, I didn't mean it that way. I meant that clearly there's complexity to it, but this complexity comes from the spec itself. Like "parsing a LALR language is straightforward", in the sense that it's a total headache, but the procedure is known and documented.

beauby commented:
@bf4 And concerning the `ActiveRecord` reference, it is there because it is the target format. At the end of the day it is just a hash, but I do not know of a better name for "hash mapping attributes to their values and `"#{key.singularize}_ids"` to arrays of ids for to-many relationships, and `"#{key}_id"` to an id for to-one relationships.

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+
+            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
+
+            fail ArgumentError, 'Invalid payload' unless payload_valid?(document)
```
I'd rather not use the exception for flow control.  I'd rather have all the code in `parse` and have it do, maybe


```ruby
if  payload_invalid?(document)
  if block_given?
    yield document
  else
    return {} # I guess this is supposed to be a benign value?
  end
end
```

and then

```ruby
InvalidDocument = Class.new(ArgumentError)
def parse!(document, options = {})
  parse(document, options) do |invalid_document)
    fail InvalidDocument, "Invalid Document: #{reason}" # reason would be good, right?
  end
end
```

beauby replied:
I am personally less comfortable with yielding stuff around for flow-control, but that's probably just me, so it's not really relevant. The way I thought of it: I defined an "unsafe" method (which would throw an exception when encountering an ill-formed document – I intended to make a custom exception with the reason and such, but haven't got there yet), and a "safe" version which would just catch the exception. I don't know whether this is common practice or specifically avoided. I'd have to have a look at how rails does stuff. In the meantime, would you mind elaborating on the reasons why using exceptions for (highly-local) flow-control does not suit you?

beauby replied:
Ping @bf4

bf4 replied:
oh, sorry... there's a number of reasons not to use exceptions for flow control, in no particular order

- they're slower than the alternative
- flow control isn't "exceptional"
- flow control isn't a good reason to possibly crash a program
- it's better to use throw/catch or blocks
  - throw/catch for unwinding the stack
  - blocks for allowing user-defined error-handling strategies

which to me I would summarize as "don't do it unless there's a really good reason"

If I had to choose over 'fail' vs. 'throw/catch', without any refactor, I'd pick throw catch, but really, I think the convention that `parse!` raises an exception is fine (if it is a good API for us), as long as the exceptional method doesn't raise an exception!

bf4 replied:
I can dig up sources if you're curious.  they're used in Nokogiri::SAX, Sinatra, and in Rails 5 callbacks https://github.com/rails/rails/pull/17227 .

bf4 replied:
Pong

B mobile phone

> On Jan 3, 2016, at 9:08 PM, Lucas Hosseini <notifications@github.com> wrote:
> 
> In lib/active_model/serializer/adapter/json_api/deserialization.rb:
> 
> > +          #     # }
> > +          #
> > +          #   parse(document, fields: [:title, :author, date: :published_at]) #=>
> > +          #     # {
> > +          #     #   title: 'Title 1',
> > +          #     #   published_at: '2015-12-20',
> > +          #     #   author_id: 2,
> > +          #     # }
> > +          #
> > +          # rubocop:disable Metrics/CyclomaticComplexity
> > +          def parse!(document, options = {})
> > +            fields = parse_fields_option(options[:fields])
> > +
> > +            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
> > +
> > +            fail ArgumentError, 'Invalid payload' unless payload_valid?(document)
> Ping @bf4
> 
> —
> Reply to this email directly or view it on GitHub.
>

ignat-zakrevsky replied:
could I intervene? Why not to use cons of both approachs?   

```ruby
DEFAULT_FALLBACK = ->(error) { fail ArgumentError, 'Invalid payload' }

def parse!(document, options = {}, &fallback)
  fallback ||= DEFAULT_FALLBACK
  
  fields = parse_fields_option(options[:fields])

  document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)

  fallback.call(document) unless payload_valid?(document)

  # ...
```

1) It will rise error as soon as possible wich is much more explicit way then simple return empty hash
2) It allows user to provide custom fallback which can be {}

ignat-zakrevsky replied:
could I intervene? Why not to use pros of both approaches?   

```ruby
DEFAULT_FALLBACK = ->(error) { fail ArgumentError, 'Invalid payload' }
SAFE_FALLBACK = ->(error) { {} }

def parse!(document, options = {}, &fallback)
  fallback ||= DEFAULT_FALLBACK
  
  fields = parse_fields_option(options[:fields])

  document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)

  fallback.call(document) unless payload_valid?(document)

  # ...


def parse(document, options = {})
  parse!(document, options, SAFE_FALLBACK)
end
```

and then call

jackturnbull replied:
Skipping back to the first post in this thread, and now that I'm starting to experiment with this PR a bit more; the current `Argument Error: Invalid Payload` format is a little too inflexible to return detailed error messages to the client. In terms of technical approach I couldn't comment but I agree with @bf4 on the following line:

```ruby
fail InvalidDocument, "Invalid Document: #{reason}" # reason would be good, right?
```

A custom error class with a reason would allow us to `rescue_from InvalidDocument, with: ...` so the default `param is missing or the value is empty: #{reason_field}` error message can be replicated, at least functionally.

I'd feel less comfortable rescuing an ArgumentError in a base controller since this is a bit of a catch all and without the reason the specific error can't be passed on to the client.

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+          #
+          # @api private
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def payload_valid?(payload)
```
Naming concern: this is the only function in the module that doesn't begin with `parse_`.  and it's usage is `unless payment_valid?` rather than `if payment_invalid?`.

Perhaps this method should be `validate_payload` then return an errors object that is `empty?` when valid.

beauby replied:
Agreed that the reason is probably a good thing to have around.

bf4 commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ Breaking changes:
 
 Features:
 
+- [#1248](https://github.com/rails-api/active_model_serializers/pull/1248) Add support for JSON API deserialization (@beauby)
```
maybe `for transforming a JSON API request document into an attributes hash`?

bf4 replied:
Also, If I'm reading this code right, this isn't integrated anywhere, so maybe it would be to mark it as an experimental feature so we can merge it and iterate faster?

beauby replied:
Yeah that would be nice!

bf4 replied:
Well, just point me to however you mark it as 'experimental, don't rely on this' in the code and changelog, and I'm ok with having it merged.  I'd generally like for us to have a way to add experimental features

bf4 replied:
(merged after squashing commits, of course)

bf4 replied:
probably would be good to create an issue output of this referencing unresolved issues, comments, what's next, etc.

beauby replied:
Added to the changelog and as a comment on the `Deserialization` module.

bf4 commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+          module_function
+
+          # Transform a JSON API document, containing a single data object, from a hash
+          # into a hash that is ready for ActiveRecord::Base.new() and such.
```
or `ActiveModelSerializers::Model.new`, etc. :).

beauby replied:
Kinda true, although the whole `blah_id` / `blah_ids` is not really handled by non-AR models, right?

bf4 replied:
I leave that up to the non-AR implementor :)

jackturnbull commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+            attributes = primary_data['attributes']
+            return false unless attributes.nil? || attributes.is_a?(Hash)
+            relationships = primary_data['relationships']
+            return false unless relationships.nil? || relationships.is_a?(Hash)
```
Feel free to correct me here (it's a bit late and I've not fully got my head around the whole project) but I think this logic isn't what we need. On https://github.com/beauby/active_model_serializers/commit/6ba0a1d77ace7da1a86f2afe0a8b8389c3d89ef9 I seem have the two following test cases:

On a hash without relationships:
```ruby
(byebug) hash = {'data'=>{'attributes'=>{'name'=>'test'}, 'type'=>'events'}}                                        
{"data"=>{"attributes"=>{"name"=>"test"}, "type"=>"events"}}

(byebug) ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(hash, fields:[:name])
*** NoMethodError Exception: undefined method `each' for nil:NilClass
```

On a hash with relationships:
```ruby
(byebug) hash = {'data'=>{'attributes'=>{'name'=>'test'}, 'type'=>'events', 'relationships' => {'author'=>{'data'=>nil}}}}
{"data"=>{"attributes"=>{"name"=>"test"}, "type"=>"events", "relationships"=>{"author"=>{"data"=>nil}}}}

(byebug) ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(hash, fields:[:name])                     
{:name=>"test"}
```

Looking at the `unless` statements on line 106 and 108, if there is no relationship key in the hash as per the first test case, the `relationships` variable will be `nil` causing the `unless` statement to pass with no return of false.

It feels like the logic may need to be separated:
* If the relationships key does not exist then continue.
* If the relationships key exists, and it is a hash then iterate.
* If the relationships key exists, and it is not a hash then return false.

beauby replied:
That's what I get for not running the tests before pushing... Thanks for spotting it @jackturnbull! Just pushed a fix.

jackturnbull replied:
Thanks again for the quick response and your work on this! I'm looking forward to getting it into production.

ignat-zakrevsky commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+          # @example
+          #   parse_attributes({ 'a' => 'b', 'c' => 'd'}, nil)
+          #    # => { :a => 'b', :c => 'd' }
+          #   parse_attributes({ 'a' => 'b', 'c' => 'd', 'e' => 'f' }, { :a, :c => :g })
```
invalid example 
```ruby
{ :a, :c => :g }
```

beauby replied:
You're right, it was meant to be `[ :a, :c => :g ]`. Thanks for spotting this.

beauby commented:
Last commit:
+ Error reporting with custom exception and reason
+ Rewrite to allow the following options:
    - only/except
    - keys (hash of translated keys, like `:author => :user`)
    - polymorphic (list of polymorphic fields (mainly relationships), for which the `#{relationship}_type` attribute should be set)

beauby commented:
Recap on the current situation:
+ Two methods are exposed (`ActiveModelSerializers::Deserialization.jsonapi_parse` and `ActiveModelSerializers::Deserialization.jsonapi_parse!`)
+ They both take the same parameters:
  - a `document` (either a Hash or an instance of `ActionController::Parameters` representing a JSON API payload)
  - an `options` hash:
    - `only` or `except`: array of white/black-listed fields (recall that in the context of JSON API, "fields" denotes both attributes and relationships)
    - `polymorphic`: array of polymorphic associations (those for which the `#{relationship}_type` must be set for `ActiveRecord` to handle it properly)
    - `keys`: hash of translated keys (for instance when the payload contains an `author` relationship which refers to an `user` association on the model)
+ The "unsafe" method (the one ending with a `!`) throws an `InvalidDocument` exception when the payload does not meet basic criteria for JSON API deserialization.
+ The "safe" method simply returns an empty hash.

There should be a bit more tests, but other than that, I believe this PR is ready. Any thoughts?

beauby commented:
Will merge once green.

beauby commented:
Ladies and gentlemen: merging!

GCorbel commented:
Woohoo! Well done @beauby !

joaomdmoura commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+          #     # }
+          #
+          #   parse(document, only: [:title, :date, :author],
+          #                   keys: { date: :published_at },
```
Nice!

joaomdmoura commented on lib/active_model/serializer/adapter/json_api/deserialization.rb:
```diff
@@ -0,0 +1,207 @@
+          #
+          #   parse(document, only: [:title, :date, :author],
+          #                   keys: { date: :published_at },
+          #                   polymorphic: [:author]) #=>
```
Not sure I got this polymorphic thing, can you help me to understand the need? :smile:

joaomdmoura replied:
nvm! got it! pretty obvious actually :grin:

maletor commented:
Should this support transforming the key *back*? Typically an application will store keys in underscore format. AMS should then convert them to dashed for JSON API. But, upon deserialization, they should be converted back to underscores. Seems like that should happen here.

bf4 commented:
@maletor Can you make a new issue? thanks
